### PR TITLE
Fixed failing tests

### PR DESCRIPTION
**Pull request by mariusvniekerk**
Msgpack tuple fix for some failing tests

### CODE REVIEW

File: distributed/protocol/core.py [hunk 1]
```diff
@@ -122,7 +122,12 @@ def loads(frames, deserialize=True, deserializers=None):
             else:
                 value = Serialized(head, fs)
 
-            get_in(key[:-1], msg)[key[-1]] = value
+            # TODO: This is problematic with tuple assignment
+            val_holder = get_in(key[:-1], msg)
+            if isinstance(val_holder, tuple):
+                msg[key[0]] = list(val_holder)
+                val_holder = get_in(key[:-1], msg)
+            val_holder[key[-1]] = value
```
Comment (mrocklin) [inline distributed/protocol/core.py]: Why this change?  Also, should we be concerned about performance impact here?

Comment (mariusvniekerk) [reply distributed/protocol/core.py]: So this is basically a result of the msgpack stuff no longer returning lists ever. 

Since tuple([1,2])[0] moans about being immutable :(.

Unfortunately this function heavily assumes that the results returned from the msgpack serialization are mutable

Comment (mrocklin) [reply distributed/protocol/core.py]: Where do we assume mutability?  Also, does this handle lists of lists of lists?  

I'm tempted to assume immutability and explicitly go mutable when necessary.  I'm running the test suite now.

Comment (mariusvniekerk) [reply distributed/protocol/core.py]: Refactored this to support lists of lists.

Comment (mariusvniekerk) [reply distributed/protocol/core.py]: Refactored this to support deeper nested things.  This does unfortunately mean a recursive `put_in` (similar to `get_in`) needs to be around

Comment (mrocklin) [reply distributed/protocol/core.py]: Sorry, I should have been more clear.  I'm inclined, if possible, to just use immutable data structures everywhere and then see where this breaks and make those specific instances explicitly mutable, by converting to lists at that location.  Do you have thoughts on this approach?

File: distributed/protocol/core.py [hunk 2]
```diff
@@ -122,7 +122,23 @@ def loads(frames, deserialize=True, deserializers=None):
             else:
                 value = Serialized(head, fs)
 
-            get_in(key[:-1], msg)[key[-1]] = value
+            from toolz import get_in
+
+            def put_in(keys, coll, val):
```
Comment (mariusvniekerk) [inline distributed/protocol/core.py:127]: Explicitly upcasts tuples as needed into lists here.

Comment (mariusvniekerk) [reply distributed/protocol/core.py:127]: Since this is in `distributed.protocol.core.loads` . (which i'm guessing is called by many many things), my assumption is that if we use choose to `loads_msgpack(use_lists=True)` here i guess all our work might be for naught unless we take this approach?

Comment (mariusvniekerk) [inline distributed/protocol/core.py:127]: Alternatively instead of this approach we could go for something like around L98

```python
msg = loads_msgpack(small_header, small_payload)
if not frames:
    return msg
```
becomes
```python
if not frames:
    msg = loads_msgpack(small_header, small_payload)
    return msg
msg = loads_msgpack(small_header, small_payload, use_list=True)
```
and then the mutability assumptions further on should hold, though i am unclear how aggresively that would change our situation.

File: distributed/protocol/core.py (hunk 1 above)
Comment (mrocklin) [reply distributed/protocol/core.py]: Ping on this.  @mariusvniekerk do you think it would work to set `use_lists=False`, use tuples everywhere, and then manually handle the hopefully few cases where we mutate a result?

File: distributed/protocol/core.py (hunk 2 above)
Comment (mrocklin) [reply distributed/protocol/core.py:127]: Ah, now I understand the need for this.  I missed the `get_in` code that this was replacing earlier.  Sorry, I was a bit dense about the need for this. 

I would be suspicious of sometimes using `use_list=True` and sometimes not.  If the `put_in` solution here works and isn't too slow then we should probably stick with it.  We might also consider returning a tuple rather than a list, again if it doesn't cause a performance hit (my guess is that it wouldn't)
